autoPyTorch/api/base_task.py

```diff
@@ -143,40 +744,100 @@ def predict(
         """
-        # TODO use the batch size and take it from the pipeline
-        # TODO argument to method for a flag that might return probabilities
-        # in case the pipeline does not have a predict_proba then continue
-        # normally and raise warning
-        return self.pipeline.predict(X_test)
+
+        # Parallelize predictions across models with n_jobs processes.
+        # Each process computes predictions in chunks of batch_size rows.
+        if self._logger is None:
+            self._logger = self._get_logger("Predict-Logger")
+
+        if self.ensemble_ is None and not self._load_models(self.resampling_strategy):
+            raise ValueError("Failed to fit an ensemble. Cannot predict")
+
+        # Mypy assert
+        assert self.ensemble_ is not None, "Load models should error out if no ensemble"
+        self.ensemble_ = cast(Union[SingleBest, EnsembleSelection], self.ensemble_)
+
+        try:
+            for i, tmp_model in enumerate(self.models_.values()):
+                if isinstance(tmp_model, (DummyRegressor, DummyClassifier)):
+                    check_is_fitted(tmp_model)
+                else:
+                    check_is_fitted(tmp_model.steps[-1][-1])
+            models = self.models_
+        except sklearn.exceptions.NotFittedError:
+            # When training a cross validation model, self.cv_models_
+            # will contain the Voting classifier/regressor product of cv
+            # self.models_ in the case of cv, contains unfitted models
+            # Raising above exception is a mechanism to detect which
+            # attribute contains the relevant models for prediction
+            try:
+                check_is_fitted(list(self.cv_models_.values())[0])
+                models = self.cv_models_
+            except sklearn.exceptions.NotFittedError:
+                raise ValueError('Found no fitted models!')
+
+        all_predictions = joblib.Parallel(n_jobs=n_jobs)(
+            joblib.delayed(_pipeline_predict)(
+                models[identifier], X_test, batch_size, self._logger, self.task_type
+            )
+            for identifier in self.ensemble_.get_selected_model_identifiers()
+        )
+
+        if len(all_predictions) == 0:
+            raise ValueError('Something went wrong generating the predictions. '
+                             'The ensemble should consist of the following '
+                             'models: %s, the following models were loaded: '
+                             '%s' % (str(list(self.ensemble_.indices_)),
+                                     str(list(self.models_))))
+
+        predictions = self.ensemble_.predict(all_predictions)
+
+        if self.task_type not in REGRESSION_TASKS:
+            # Make sure average prediction probabilities
+            # are within a valid range
+            # Individual models are checked in _model_predict
+            predictions = np.clip(predictions, 0.0, 1.0)
```

Comment:
This could be dangerous if there are errors in the voter and make debugging hard. Maybe better throw an exception or does this actually occur legally in some cases?

Reply:
It was happening for some of the runs, but that was in individual pipelines where an error is raised in _pipeline_predict. But yeah I think raising an error might be better, so I'll do that here as well.